### ISSUE TITLE
Failed to sync the nodeGraph when starting a node.

### ISSUE THREAD

**Issue by tegefaulkes** [MatrixAI/Polykey-CLI#104]
### Describe the bug

The details on this one is a bit light and the problem needs some exploration. 

While demonstrating `Polykey-CLI` the node failed to failed to sync the node graph. On top of this, any attempt to ping the seed nodes failed. Now the only way `SyncNodeGraph` should fail is if it fails to make connections to any seed nodes. This is very odd since if the seed nodes are working then no node should really fail to make a connection. It seems more likely that the seed nodes are down currently for whatever reason.

### Expected behaviour

Running a node should not fail to sync the node graph and pings should succeed.

**Comment by tegefaulkes** [comment]
This is an odd one. The only thing I really know about this is that the
1.  syncing failed.
2. Could not ping the seed nodes. This would cause the failure to sync.
3. Was done on a mobile network.

So far as I can tell it would only be cause by the seed nodes being down. So to look into this I'd do the following. 

1. Sanity check that the seed nodes are running fine In AWS.
2. Sanity check by running a node
3. Failing to achieve find any failure then we'd have to mothball the problem until we can find more details or recreate somehow.

**Comment by CMCDragonkai** [comment]
They are not down. They are live on the PKNS on the mainnet, and normal usage is fine. However he was on mobile network though. So it's something. Plus we could see the node connections, we just couldn't ping them. I suspect the initial node connections came from DNS.

**Comment by CMCDragonkai** [comment]
@amydevs

**Comment by tegefaulkes** [comment]
I'm going do some manual testing on this now. Starting with sanity checking and then testing a mobile network.

**Comment by tegefaulkes** [comment]
Sanity check shows I'm establishing connections to seed nodes.

```
[nix-shell:~/workspace/Polykey-CLI]$ npm run polykey -- -np tmp/nodeasdasd nodes connections

> polykey-cli@0.1.3 polykey
> ts-node src/polykey.ts -np tmp/nodeasdasd nodes connections

host            hostname        nodeIdEncoded                                           port    timeoutusageCount
13.54.214.222   N/A             v75epnbm856v0q8qdau6vs45675hs0b8q74n3nup9mvkln7rd7u10   1314    115499 0
3.131.149.71    N/A             v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g   1314    143567 0
```

Testing using internet tethered from my phone works just fine as well.

```
[nix-shell:~/workspace/Polykey-CLI]$ npm run polykey -- -np tmp/nodeasdasd nodes connections

> polykey-cli@0.1.3 polykey
> ts-node src/polykey.ts -np tmp/nodeasdasd nodes connections

host            hostname        nodeIdEncoded                                           port    timeout usageCount
13.54.214.222   N/A             v75epnbm856v0q8qdau6vs45675hs0b8q74n3nup9mvkln7rd7u10   1314    112440  0
3.131.149.71    N/A             v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g   1314    140928  0
```

**Comment by CMCDragonkai** [comment]
Ok I expected that it should work since these are public nodes anyway.

**Comment by CMCDragonkai** [comment]
Just need a `nodes ping` command result and then I guess we can close this for now.

**Comment by tegefaulkes** [comment]
Pinging also works. 

```
[nix-shell:~/workspace/Polykey-CLI]$ npm run polykey -- -np tmp/nodeasdasd nodes ping v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g

> polykey-cli@0.1.3 polykey
> ts-node src/polykey.ts -np tmp/nodeasdasd nodes ping v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g

Node is Active.

```

and on mobile.

```
[nix-shell:~/workspace/Polykey-CLI]$ npm run polykey -- -np tmp/nodeasdasd nodes ping v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g

> polykey-cli@0.1.3 polykey
> ts-node src/polykey.ts -np tmp/nodeasdasd nodes ping v13o9h7aahij42b3b4lt0lu1ovlqp46t2sb1l1amk1onumdor4v1g

Node is Active.

```

**Comment by CMCDragonkai** [comment]
This is still a problem.

George's Wireshark shows no udp packets being sent on 1314 nor packets being sent to the mainnet node IPs.

It's not a problem of sending UDP cause it's possible to force it with `nc -u`. And then Wireshark shows it.

But for some reason it's not sending any UDP packets to the mainnet nodes.

@tegefaulkes can you retest all the above commands but over a mobile network (make sure to disconnect from wired) and run Wireshark with the filter `udp.port == 1314`.

**Comment by CMCDragonkai** [comment]
![IMG20240203194324.jpg](https://github.com/MatrixAI/Polykey-CLI/assets/640797/718288e4-bac5-4b77-a9a5-f6945252ee2d)

![IMG20240203194402.jpg](https://github.com/MatrixAI/Polykey-CLI/assets/640797/52b6c328-6134-40ba-bbc3-97eea49575ea)

There is not enough information coming out of the logs even at debug level to explain why there's no packets being sent at all. Wireshark basically doesn't detect any UDP packets being sent to the mainnet nodes or being received to the agent port.

@tegefaulkes extra information needs to be printed indicating why this is happening.

**Comment by CMCDragonkai** [comment]
Also changing wifi networks causes the agent to freeze and no longer respond to SIGINT.

**Comment by CMCDragonkai** [comment]
It has something to do with the OS DNS settings. By using dnsmasq the os couldn't resolve DNS SD records I think. This should show up as an error on Polykey if this is an issue.

It can worked around by using our own DNS resolution in-app.

**Comment by CMCDragonkai** [comment]
1. We need error reporting for cases where DNS resolution fails and causes failed connection setup.
2. We need to deal with wifi drop outs or switches/disconnects.
3. We should have a fallback to in-app DNS resolution if necessary.

**Comment by tegefaulkes** [comment]
DNS should be working fine. Otherwise you wouldn't have any nodes in the nodeGraph when listing them with `nodes getall`. It's *possible* there's a weird but where it resolves the nodes but attempts no connections when syncing the node graph.

But I can see connection attempts in the logs above, It's a bit hard to follow though with all the DB transaction logs. Posting full logs and the wireshark logs here would help a lot with debugging.

As for testing on a mobile network, I already have and made sure that the wireless/wired was disconnected. I only had USB tethering from my phone. I can also test with just wifi tethering as well.

I could expand the logging a bit. For now it will log out attempts, successes or failures but not each packet. Logging for each packing could affect performance pretty badly given how low level it is.

**Comment by CMCDragonkai** [comment]
It only worked after the dnsmasq was switched off. So OS DNS can fail to resolve DNS SD. The main issue here is no reports on its failure to resolve the DNS to ips which is why no udp packets was sent.

5 Feb 2024 11:31:06 Brian Botha ***@***.***>:

**Comment by CMCDragonkai** [comment]
Mobile network was not the problem. It was about DNS. DNS has 2 parts 1. Getting the bootstrap list, and 2. Resolving to the specific IP. The second part is what didn't occur without any messages. Remember no UDP packets were sent while dnsmasq was enabled.

5 Feb 2024 11:31:06 Brian Botha ***@***.***>:

**Comment by CMCDragonkai** [comment]
> Also changing wifi networks causes the agent to freeze and no longer respond to SIGINT.

Not good for UX, forces user to send SIGKILL.

**Comment by tegefaulkes** [comment]
I'm aware that loosing the network by switching off wifi currently crashes. Having it freeze when switching is a weird one but I can look into it while resolving the other thing.

I can run a test to see if failing to resolve/hanging DNS query will cause this problem.

**Comment by tegefaulkes** [comment]
We concluded that this was DNS related and not a problem inside Polykey itself right? Is this still a problem or can I close it @CMCDragonkai ?

**Comment by CMCDragonkai** [comment]
> 1. We need error reporting for cases where DNS resolution fails and causes failed connection setup.
> 2. We need to deal with wifi drop outs or switches/disconnects.
> 3. We should have a fallback to in-app DNS resolution if necessary.

This still needs to be done. Create new issues for these before we close this investigation.

**Comment by tegefaulkes** [comment]
Ok, I'll spec out new issues and close this one out.

**Comment by tegefaulkes** [comment]
> > 1. We need error reporting for cases where DNS resolution fails and causes failed connection setup.
> > 2. We need to deal with wifi drop outs or switches/disconnects.
> > 3. We should have a fallback to in-app DNS resolution if necessary.
> 
> This still needs to be done. Create new issues for these before we close this investigation.

For 1. and 3. - https://github.com/MatrixAI/Polykey/issues/729
For 2. - #115
